### PR TITLE
Add LIP "Add weights to Lisk-BFT consensus protocol"

### PR DESCRIPTION
**Pull request by janhack**

### CODE REVIEW

File: proposals/lip-add_weights_to_lisk-bft_consensus_protocol.md [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+This LIP defines a part of the state transition logic of the BFT module introduced in the [BFT module LIP][bft-module-lip].
+The LIP therefore requires a hard fork as introducing the BFT module implies a hardfork.
+
+[bft-module-lip]: https://research.lisk.com/t/introduce-bft-module/321
+[lip-34]: https://github.com/LiskHQ/lips/blob/master/proposals/lip-0034.md
+[lisk-bft-lip]: https://github.com/LiskHQ/lips/blob/master/proposals/lip-0014.md
+[lisk-bft-lip-comp]: https://github.com/LiskHQ/lips/blob/master/proposals/lip-0014.md#computing-prevotes-and-precommits
+[lisk-bft-paper]: https://arxiv.org/abs/1903.11434
+[poa-lip]: https://research.lisk.com/t/proof-of-authority-validator-selection-mechanism/288
+[validator-module-lip]: https://research.lisk.com/t/introduce-validators-module/317
```
Comment (ricott1) [inline proposals/lip-add_weights_to_lisk-bft_consensus_protocol.md:314]:
```suggestion
[validator-module-lip]: https://github.com/LiskHQ/lips/blob/master/proposals/lip-0044.md
```

File: proposals/lip-add_weights_to_lisk-bft_consensus_protocol.md [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+
+This LIP defines a part of the state transition logic of the BFT module introduced in the [BFT module LIP][bft-module-lip].
+The LIP therefore requires a hard fork as introducing the BFT module implies a hardfork.
+
+[bft-module-lip]: https://research.lisk.com/t/introduce-bft-module/321
+[lip-34]: https://github.com/LiskHQ/lips/blob/master/proposals/lip-0034.md
+[lisk-bft-lip]: https://github.com/LiskHQ/lips/blob/master/proposals/lip-0014.md
+[lisk-bft-lip-comp]: https://github.com/LiskHQ/lips/blob/master/proposals/lip-0014.md#computing-prevotes-and-precommits
+[lisk-bft-paper]: https://arxiv.org/abs/1903.11434
+[poa-lip]: https://research.lisk.com/t/proof-of-authority-validator-selection-mechanism/288
```
Comment (ricott1) [inline proposals/lip-add_weights_to_lisk-bft_consensus_protocol.md:313]:
```suggestion
[poa-lip]: https://github.com/LiskHQ/lips/blob/master/proposals/lip-0047.md
```

File: proposals/lip-add_weights_to_lisk-bft_consensus_protocol.md [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+The decision threshold `τ` in the paper corresponds to the precommit threshold used in this LIP up to rescaling and the subtle difference that the weight of precommits in the paper has to be strictly more than `τ` but in these specifications it only has to be greater or equal than `precommitThreshold(h)` (as we restrict the range of `precommitThreshold(h)` such that `precommitThreshold(h) / aggregateBFTWeight(h) > 1/3`).
+Similarly, we choose `prevoteThreshold(h)` such that `prevoteThreshold(h) / aggregateBFTWeight(h) > 2/3`.
+
+The parameter `ρ`, which determines how many blocks a block can imply prevotes for, is given by `3*LSK_BFT_BATCH_SIZE` in the specifications in this LIP, as also in LIP 0014.
+Here `LSK_BFT_BATCH_SIZE` is a constant of the  [BFT module LIP][bft-module-lip].
+As the round length in a PoA chain can vary, we require that the maximum possible round length in a PoA chain, given by the constant `MAX_NUM_VALIDATORS` defined in the [Proof-of-Authority LIP][poa-lip], is always at most the constant `LSK_BFT_BATCH_SIZE`.
```
Comment (ricott1) [inline proposals/lip-add_weights_to_lisk-bft_consensus_protocol.md:92]:
```suggestion
As the round length in a PoA chain can vary, we require that the maximum possible round length in a PoA chain, given by the constant `MAX_NUM_VALIDATORS` defined in [LIP 0047][poa-lip], is always at most the constant `LSK_BFT_BATCH_SIZE`.
```

**Review by ricott1 — APPROVED** [review]
Minor comments about updating some links

File: proposals/lip-add_weights_to_lisk-bft_consensus_protocol.md [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+In these blockchains, there is a fixed number of active delegates, i.e., those delegates participating in finalizing blocks, and all of them have equal weight.
+One exception to that is the bootstrap period introduced as part of the new genesis block format in [LIP 0034][lip-34].
+During the bootstrap period none of the bootstrap delegates contributes to finalizing blocks, so that blocks can only be finalized after the bootstrap period is over.
+
+For more flexibility and to also support other validator selection mechanisms, we want to allow for different BFT weights of the active validators, i.e., different weights attributed to the prevotes and precommits cast by a validator.
+These weights and the threshold for finalizing a block should also not be constant, but should be allowed to change over time. This is, in particular, required in conjunction with the Proof-of-Authority validator selection mechanism introduced in the [PoA][module] where validators can be added and removed or their weights changed with a simple transaction and without a hard fork.
```
Comment (MaximeGagnebin) [inline proposals/lip-add_weights_to_lisk-bft_consensus_protocol.md:30]: This link to the PoA module seems broken.

**Review by MaximeGagnebin — APPROVED** [review]
Just a small typo style comment. Good job!

File: proposals/lip-add_weights_to_lisk-bft_consensus_protocol.md (hunk 2 above)
Comment (janhack) [reply proposals/lip-add_weights_to_lisk-bft_consensus_protocol.md:319]: Fixed.

File: proposals/lip-add_weights_to_lisk-bft_consensus_protocol.md (hunk 1 above)
Comment (janhack) [reply proposals/lip-add_weights_to_lisk-bft_consensus_protocol.md:320]: Fixed.

File: proposals/lip-add_weights_to_lisk-bft_consensus_protocol.md (hunk 3 above)
Comment (janhack) [reply proposals/lip-add_weights_to_lisk-bft_consensus_protocol.md:92]: changed to "defined in the PoA module" instead

File: proposals/lip-add_weights_to_lisk-bft_consensus_protocol.md (hunk 4 above)
Comment (janhack) [reply proposals/lip-add_weights_to_lisk-bft_consensus_protocol.md]: I fixed the link.

**Comment by janhack** [comment]
I addressed the review comments and fixed one issue that I noticed from the implementation. Please check again.

**Review by ricott1 — APPROVED** [review]
LGTM, great job!

**Review by MaximeGagnebin — APPROVED** [review]
LGTM

**Review by mitsuaki-u — APPROVED** [review]
LGTM!

**Review by shuse2 — APPROVED** [review]

**Review by karmacoma — APPROVED** [review]

**Pull request merged**